src/orchard/xref.clj

```diff
@@ -13,21 +32,87 @@
-  {:added "0.5"}
-  [val]
-  (let [val (as-val val)]
-    (set (some->> val class .getDeclaredFields
+(defn- f->sym [^clojure.lang.AFn f]
+  (-> f .getClass .getName repl/demunge symbol))
+
+(defn- fn-source [f]
+  (hunt-down-source (f->sym f)))
+
+(defn- fn-name [^java.lang.Class f]
+  (-> f .getName repl/demunge symbol))
+
+(defonce classbytes (atom {}))
+
+(defn- recompile [ns-sym form]
+  (push-thread-bindings
+   {clojure.lang.Compiler/LOADER
+    (proxy [clojure.lang.DynamicClassLoader] [@clojure.lang.Compiler/LOADER]
+      (defineClass
+        ([name bytes src]
+         (swap! classbytes assoc name bytes)
+         (proxy-super defineClass name bytes src))))})
+  (try
+    (let [line @clojure.lang.Compiler/LINE
+          column @clojure.lang.Compiler/COLUMN
+          line (if-let [line (:line (meta form))]
+                 line
+                 line)
+          column (if-let [column (:column (meta form))]
+                   column
+                   column)]
+      (push-thread-bindings {clojure.lang.Compiler/LINE line
+                             clojure.lang.Compiler/COLUMN column})
+      (try
+        (let [form (macroexpand form)]
+          (when (and (coll? form) (= 'clojure.core/fn (first (nth form 2 nil))))
+            (binding [*ns* (create-ns ns-sym)]
+              (clojure.lang.Compiler/analyze
+               clojure.lang.Compiler$C/EVAL
+               (nth form 2)))))
+        (finally
+          (pop-thread-bindings))))
+    (finally
+      (pop-thread-bindings))))
+
+(defn- fn-deps-class
+  [v]
+  (let [^java.lang.Class v (if (class? v)
+                             v
+                             (eval v))]
+    (set (some->> v .getDeclaredFields
                   (keep (fn [^java.lang.reflect.Field f]
                           (or (and (identical? clojure.lang.Var (.getType f))
                                    (java.lang.reflect.Modifier/isPublic (.getModifiers f))
                                    (java.lang.reflect.Modifier/isStatic (.getModifiers f))
                                    (-> f .getName (.startsWith "const__"))
-                                   (.get f val))
+                                   (.get f (fn-name v)))
                               nil)))))))
 
+(defn fn-deps-compiling
+  "Returns a set with all the functions invoked by `val`.
+  `val` can be a function value, a var or a symbol."
+  {:added "0.5"}
+  [f]
+  (reset! classbytes {})
+  (let [f (as-val f)]
+    (when-let [source (and (fn? f) (fn-source f))]
+      (let [sym (f->sym f)
+            b (do (recompile (-> sym namespace symbol) source) @classbytes)
+            class-names (map first b)
+            deps (set (mapcat #(-> % symbol fn-deps-class) class-names))]
+        deps))))
+
+(defn fn-deps [s]
+  (when-let [^clojure.lang.AFn v (as-val s)]
+    (let [f-class-name (-> v .getClass .getName)
+          field (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
```

Comment:
can you please get the field by name? so that we can see what field it is

Reply:
that's a problem I'm still having, if I change that line to (.getDeclaredField clojure.lang.DynamicClassLoader "classCache") which is basically the same thing, I get an IllegalAccessException on the next line.
The same thing happens if I type hint field to ^java.lang.reflect.Field

Reply:
At least `field` could be named more expressively